### PR TITLE
Function pointer bug fixes

### PR DESCRIPTION
**Pull request by 333fred**
Fixes a couple of function pointer bugs:

* Commit 1 ensures that we use translated local function symbols, not the original symbol, when emitting the `ldftn` instruction, and corrects handling of non-static local functions. This fixes https://github.com/dotnet/roslyn/issues/45447.
* Commit 2 ensures that we correctly bind arguments for error recovery in all cases. Fixes https://github.com/dotnet/roslyn/issues/45418.
* Commit 5 ensures that we don't emit an ldftn the result is not used.

@dotnet/roslyn-compiler @AlekseyTs for review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 1]
```diff
@@ -1680,15 +1681,15 @@ private BoundFunctionPointerInvocation BindFunctionPointerInvocation(SyntaxNode
...
+                    methods,
                     typeContainingConstructor: null,
                     delegateTypeBeingInvoked: null,
                     returnRefKind: funcPtr.Signature.RefKind);
 
                 return new BoundFunctionPointerInvocation(
                     node,
                     boundExpression,
-                    analyzedArguments.Arguments.SelectAsArray((expr, args) => args.binder.BindToNaturalType(expr, args.diagnostics), (binder: this, diagnostics)),
+                    BuildArgumentsForErrorRecovery(analyzedArguments, StaticCast<MethodSymbol>.From(methods)),
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1692]: Can you elaborate a bit on why this change is beneficial? Asking for my education here.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1692]: This forces things like UnboundLambdas to be bound for error recovery, and also forces any variables pending inference to be inferred as best they can. This last bit was the cause of the crash.

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs [hunk 2]
```diff
@@ -1303,6 +1303,30 @@ public override BoundNode VisitDelegateCreationExpression(BoundDelegateCreationE
...
+
+                RemapLocalFunction(
+                    node.Syntax,
+                    node.TargetMethod,
+                    out var receiver,
+                    out var remappedMethod,
+                    ref arguments,
+                    ref argRefKinds);
+
+                Debug.Assert(arguments.IsDefault && argRefKinds.IsDefault && receiver.Kind == BoundKind.TypeExpression);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1322]: >receiver.Kind == BoundKind.TypeExpression [](start = 77, length = 41)

What is this type?

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs [hunk 3]
```diff
@@ -1303,6 +1303,30 @@ public override BoundNode VisitDelegateCreationExpression(BoundDelegateCreationE
             return base.VisitDelegateCreationExpression(node);
         }
 
+        public override BoundNode VisitFunctionPointerLoad(BoundFunctionPointerLoad node)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1306]: >VisitFunctionPointerLoad [](start = 34, length = 24)

Did you check other places where we override VisitDelegateCreationExpression to confirm that we don't need to add VisitFunctionPointerLoad?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs [hunk 4]
```diff
@@ -7101,6 +7101,199 @@ .locals init (delegate*<System.Span<char>, System.Span<char>> V_0)
...
+        {
+            var verifier = CompileAndVerifyFunctionPointers(@"
+unsafe class FunctionPointer
+{
+    public static void Main()
+    {
+        delegate*<void> a = &local;
+        a();
+
+        static void local() => System.Console.Write(""local"");
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:7115]: Consider adding a test where the contents of this method body exist within a lambda expression. Verifying the rewriting works in a nested environment.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3)

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1306]: I've taken a look. I'll add an override in `MethodToClassRewriter`, but I can't find a case that it will actually change, as other than this it's only used for iterator and async method rewriting. Since we don't allow you to refer to the compiler-generated underlying methods of these from source, there's nothing that we can test. If we add a new thing that gets remapped in the future, though, it will function as expected.

---
In reply to: [445956108](https://github.com/dotnet/roslyn/pull/45473#discussion_r445956108) [](ancestors = 445956108)

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1322]: Made the `var` explicit and added some more verification of the remapped method.

---
In reply to: [445955503](https://github.com/dotnet/roslyn/pull/45473#discussion_r445955503) [](ancestors = 445955503)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

This PR now addresses an additional issue I found when testing locally, where an `ldftn` would be left on the stack when the result was unused. @jaredpar, I'd like another review with that additional fix.

---
In reply to: [649949098](https://github.com/dotnet/roslyn/pull/45473#issuecomment-649949098) [](ancestors = 649949098)

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs [hunk 5]
```diff
@@ -3466,11 +3466,15 @@ private void EmitCallCleanup(SyntaxNode syntax, UseKind useKind, MethodSymbol me
...
+        private void EmitLoadFunction(BoundFunctionPointerLoad load, bool used)
         {
             Debug.Assert(load.Type is { TypeKind: TypeKind.FunctionPointer });
-            _builder.EmitOpCode(ILOpCode.Ldftn);
-            EmitSymbolToken(load.TargetMethod, load.Syntax, optArgList: null);
+
+            if (used)
+            {
+                _builder.EmitOpCode(ILOpCode.Ldftn);
+                EmitSymbolToken(load.TargetMethod, load.Syntax, optArgList: null);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:3476]: >EmitSymbolToken(load.TargetMethod, load.Syntax, optArgList: null); [](start = 16, length = 66)

Can this have a side effect, for example an exception? Is this conditional handling consistent with what we do for delegate creation? I.e. if a delegate instance is not used, are we not loading the address of a target method?

Comment (333fred) [reply src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:3476]: If I hadn't fixed the remapped local function bug, it would have an effect, but the only effect from this I could see now is if a tree-trimming algorithm decides the referenced method isn't used (which it isn't) and decides to remove it. However, when I implemented this I checked our handling for delegates:

```cs
static void M()
{
    System.Action a = M;
}
```

produces this IL in release mode:

```
    .method private hidebysig static 
        void M () cil managed 
    {
        // Method begins at RVA 0x2050
        // Code size 1 (0x1)
        .maxstack 8

        IL_0000: ret
    } // end of method C::M
```

---
In reply to: [446421466](https://github.com/dotnet/roslyn/pull/45473#discussion_r446421466) [](ancestors = 446421466)

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:3476]: My intuition was that the tree trimming aspect was fine, and likely desired. The fact that it lines up with `delegate` behavior makes me feel good about this outcome.

File: src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs [hunk 6]
```diff
@@ -506,6 +506,11 @@ public override BoundNode VisitDelegateCreationExpression(BoundDelegateCreationE
             return node.Update(rewrittenArgument, method, node.IsExtensionMethod, type);
         }
 
+        public override BoundNode VisitFunctionPointerLoad(BoundFunctionPointerLoad node)
+        {
+            return node.Update(VisitMethodSymbol(node.TargetMethod), node.Type);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs:511]: >node.Type [](start = 69, length = 9)

It looks like we should also visit type.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 5)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [650448547](https://github.com/dotnet/roslyn/pull/45473#issuecomment-650448547) [](ancestors = 650448547)
